scoring_engine/web/views/api/flags.py

```diff
@@ -51,32 +50,40 @@ def api_flags():
 @login_required
 @cache.cached(make_cache_key=make_cache_key)
 def api_flags_solves():
-    # team = session.query(Team).get(current_user.team.id)
-    # if team is None or not current_user.team == team or not (current_user.is_red_team or current_user.is_white_team):
-    #     return jsonify({"status": "Unauthorized"}), 403
-
-    # Build a dynamic case query for each team
-    columns = [Flag.id.label("flag_id")]
+    if not current_user.is_red_team and not current_user.is_white_team:
+        return jsonify({"status": "Unauthorized"}), 403
 
     # Get all flags and teams
-    # all_flags = session.query(Flag.id).all()
-    all_teams = Team.get_all_blue_teams()
-
-    for team in all_teams:
-        columns.append(func.max(case((Solve.team_id == team.id, True), else_=False)).label(f"{team.name}"))
-
-    # Query to pivot the data
     now = datetime.utcnow()
     early = now + timedelta(minutes=int(Setting.get_setting("agent_show_flag_early_mins").value))
-    results = (
-        session.query(*columns)
-        .order_by(Flag.start_time)
-        .outerjoin(Solve, Solve.flag_id == Flag.id)
-        .filter(and_(early > Flag.start_time, now < Flag.end_time))
-        .group_by(Flag.id)
-        .all()
-    )
-
-    data = [tuple(row) for row in results]
-
-    return jsonify(data=data)
+    active_flags = session.query(Flag).filter(and_(early > Flag.start_time, now < Flag.end_time)).order_by(Flag.start_time).all()
+    active_flag_ids = [flag.id for flag in active_flags]
+
+    # Flag Solve Status
+    all_hosts = session.query(Service.name.label("service_name"), Service.port, Service.team_id, Service.host, Team.name.label("team_name"), func.coalesce(Solve.id, None).label("solve_id"), func.coalesce(Flag.id, None).label("flag_id"), func.coalesce(Flag.perm, None).label("flag_perm"), func.coalesce(Flag.platform, None).label("flag_platform")).filter(Service.check_name == "AgentCheck").outerjoin(Solve, and_(Solve.host == Service.host, Solve.team_id == Service.team_id)).filter(or_(Solve.flag_id.in_(active_flag_ids), Solve.host.is_(None))).join(Flag, or_(Flag.id == Solve.flag_id, Solve.flag_id == None)).join(Team, Team.id == Service.team_id).order_by(Service.name, Service.team_id).all()
```

Comment:
this query is a monster...i think i have a way to improve it. However, it does appear to be relatively performant right now. Still investigating if there is an active bug causing the table to not display in certain scenarios